src/effects/native/tremoloeffect.cpp

```diff
@@ -0,0 +1,224 @@
+    "0 - 1 period shift");
+    phase->setControlHint(
+        EffectManifestParameter::ControlHint::KNOB_LINEAR);
+    phase->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    phase->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
+    phase->setDefault(0);
+    phase->setMinimum(0);
+    phase->setMaximum(1);
+
+    EffectManifestParameter* quantize = manifest.addParameter();
+    quantize->setId("quantize");
+    quantize->setName("Quantize");
+    quantize->setShortName("Quantize");
+    quantize->setDescription(
+        "Round the Rate parameter to the nearest whole division of a beat.");
+    quantize->setControlHint(
+        EffectManifestParameter::ControlHint::TOGGLE_STEPPING);
+    quantize->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    quantize->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
+    quantize->setDefault(1);
+    quantize->setMinimum(0);
+    quantize->setMaximum(1);
+
+    EffectManifestParameter* triplet = manifest.addParameter();
+    triplet->setId("triplet");
+    triplet->setName("Triplets");
+    triplet->setShortName(QObject::tr("Triplet"));
+    triplet->setDescription("When the Quantize parameter is enabled, divide "
+                            "rounded 1/4 beats of Rate parameter by 3.");
+    triplet->setControlHint(
+        EffectManifestParameter::ControlHint::TOGGLE_STEPPING);
+    triplet->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    triplet->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
+    triplet->setDefault(0);
+    triplet->setMinimum(0);
+    triplet->setMaximum(1);
+
+    return manifest;
+}
+
+TremoloEffect::TremoloEffect(EngineEffect* pEffect,
+                             const EffectManifest& manifest)
+        : m_pRateParameter(pEffect->getParameterById("rate")),
+          m_pShapeParameter(pEffect->getParameterById("shape")),
+          m_pSmoothParameter(pEffect->getParameterById("smooth")),
+          m_pPhaseParameter(pEffect->getParameterById("phase")),
+          m_pQuantizeParameter(pEffect->getParameterById("quantize")),
+          m_pTripletParameter(pEffect->getParameterById("triplet")) {
+    Q_UNUSED(manifest);
+}
+
+TremoloEffect::~TremoloEffect() {
+}
+
+void TremoloEffect::processChannel(const ChannelHandle& handle,
+                                TremoloGroupState* pState,
+                                const CSAMPLE* pInput, CSAMPLE* pOutput,
+                                const unsigned int numSamples,
+                                const unsigned int sampleRate,
+                                const EffectProcessor::EnableState enableState,
+                                const GroupFeatureState& groupFeatures) {
+    Q_UNUSED(handle);
+
+    const double shape  = m_pShapeParameter->value();
+    const double smooth = m_pSmoothParameter->value();
+
+    unsigned int currentFrame = pState->currentFrame;
+    double       gain         = pState->gain;
```

Comment:
nit: we do not care about alignment of code lines.
This will likely brake if you refactor tho code.
Use just single spaces.